### PR TITLE
Limit the damage when the temporary directory with qgd file is deleted

### PR DESCRIPTION
**Pull request by pblottiere**
## Description

There's a lot of issues about the *Unable to save auxiliary storage* message (see https://github.com/qgis/QGIS/issues/26365 for example). Actually, this error occurs when the temporary directory where the .qgd file is stored is deleted (probably by a third party component like some kind of "cleaning" tool).

In this case, last changes on the project are lost forever and it may be very frustrating. I don't know how to prevent the temporary directory from being deleted, but I propose in this PR to limit the damage by:

1. saving the qgs project in the qgz archive even if the qgd file doesn't exist anymore 
2. saving the last version of the qgd file in the qgz archive
3. auto-fix the auxiliary storage mechanism (reset archive and auxiliary storage instances)

This way, only the last changes on the auxiliary file are lost. I don't really know how to reproduce the error while a simple `rm -rf /tmp/QGIS3*` does the trick...

Good idea or bad idea? Opinions are welcome :).

TODO:
- [x] don't create the qgd file by default when saving in qgz
- [x] add some tests

### CODE REVIEW

File: src/core/project/qgsproject.cpp [hunk 1]
```diff
@@ -3338,25 +3338,50 @@ bool QgsProject::zip( const QString &filename )
   const QFileInfo info( qgsFile );
   const QString asFileName = info.path() + QDir::separator() + info.completeBaseName() + "." + QgsAuxiliaryStorage::extension();
 
+  bool asOk = true;
```
Comment (nyalldawson) [inline src/core/project/qgsproject.cpp:3341]:
```suggestion
  bool auxiliaryStorageSavedOk = true;
```

File: src/core/project/qgsproject.cpp [hunk 2]
```diff
@@ -3338,25 +3338,50 @@ bool QgsProject::zip( const QString &filename )
...
   if ( ! saveAuxiliaryStorage( asFileName ) )
   {
     const QString err = mAuxiliaryStorage->errorString();
-    setError( tr( "Unable to save auxiliary storage ('%1')" ).arg( err ) );
-    return false;
+    setError( tr( "Unable to save auxiliary storage file ('%1'). The project is still saved but last changes on auxiliary data are lost. It is recommended to reload the project." ).arg( err ) );
```
Comment (nyalldawson) [inline src/core/project/qgsproject.cpp:3345]:
```suggestion
    setError( tr( "Unable to save auxiliary storage file ('%1'). The project has been saved but the latest changes to auxiliary data cannot be recovered. It is recommended to reload the project." ).arg( err ) );
```

File: src/core/project/qgsproject.cpp [hunk 3]
```diff
@@ -3338,25 +3338,50 @@ bool QgsProject::zip( const QString &filename )
...
+    asOk = false;
+
+    // fixes the current archive and keep the previous version of qgd
+    if ( !mArchive->exists() )
+    {
+      mArchive.reset( new QgsProjectArchive() );
+      mArchive->unzip( mFile.fileName() );
+      mArchive->clearProjectFile();
+
+      const QString asFile = mArchive->auxiliaryStorageFile();
```
Comment (nyalldawson) [inline src/core/project/qgsproject.cpp:3355]:
```suggestion
      const QString auxiliaryStorageFile = mArchive->auxiliaryStorageFile();
```

**Comment by nyalldawson** [comment]
> Good idea or bad idea? Opinions are welcome :).

I can't think of a better approach! :smile:

**Comment by pblottiere** [comment]
>> Good idea or bad idea? Opinions are welcome :).

> I can't think of a better approach! 

@nyalldawson thanks for the review :)

**Comment by pblottiere** [comment]
Hello @nyalldawson,

Could you take a look on this PR please? I'd love to see this coming in the next release due to the high rate of complaints...

Or @elpaso maybe?

File: tests/src/python/test_qgsauxiliarystorage.py [hunk 4]
```diff
@@ -556,6 +557,49 @@ def testInvalidPrimaryKey(self):
...
+        path = tmpPath()
+        qgz = path + '.qgz'
+        self.assertTrue(p.write(qgz))
+        self.assertTrue(os.path.exists(qgz))
+
+        # Check the content of the archive: auxiliary database doesn't exist
+        # because it's empty
+        archive = QgsProjectArchive()
+        archive.unzip(qgz)
+        self.assertTrue(archive.auxiliaryStorageFile())
```
Comment (elpaso) [inline tests/src/python/test_qgsauxiliarystorage.py:601]: Can you use notEqual instead of using a bool to check for an empty string?

**Review by elpaso — APPROVED** [review]

**Comment by Houska1** [comment]
Thanks for doing this. Looking forward to seeing this issue go away, hopefully.

Your explanation that something (on Windows, seems to be a Win problem?) deletes the temp directory as cleanup makes sense. The times I've experienced it, and have seen others report it, all involved leaving QGIS running in the background with a project open, and sleep/hibernation of the system or a long period of inactivity, which might prompt some sort of OS cleanup.
I'm not sure, but all instances I know of have occurred when the .qgd file *is* actually empty, so presumably the cleanup process is feeling entitled to do its thing by seeing a temp subdir whose only contents is a zero-length file presumably not in use by any process. 

So in case people continue to report this (now as "I got a weird warning about auxiliary files, not sure what they are"...) might then make sense to fix up silently in case there is no aux info to save anyway.

**Comment by pblottiere** [comment]
@Houska1 Thanks a lot for your constructive feedback (here and on the corresponding issue), it's well appreciated :+1:. Unfortunately it hasn't been merged in time for the 3.20 release and it may be to risky for a backport... I'll take a look.

**Comment by Houska1** [comment]
@pblottiere No worries, I just saw it go by merged into master, so assume it will appear in 3.22, and maybe worth it for 3.20.1? 
BTW, my supposition (zero length qgd) may be off, and even if it's right, this is already friendlier. So can probably leave it for now as you have it, just be prepared for it to migrate into "I got this weird warning" bug reports :)   

Or maybe just make the error message in qgsproject.cpp a bit more user-friendly for less advanced users, something like:

"Unable to save auxiliary storage file ('%1'). The project has been saved but the latest changes to auxiliary data cannot be recovered. It is recommended to reload the project. You can ignore this message if auxiliary storage is not being used."

**Comment by pblottiere** [comment]
> So can probably leave it for now as you have it, just be prepared for it to migrate into "I got this weird warning" bug reports :)

> Or maybe just make the error message in qgsproject.cpp a bit more user-friendly for less advanced users, something like:

> "Unable to save auxiliary storage file ('%1'). The project has been saved but the latest changes to auxiliary data cannot be recovered. It is recommended to reload the project. You can ignore this message if auxiliary storage is not being used."

Actually I also removed the automatic creation of the qgd file. So the underlying issue shouldn't happen anymore if auxiliary storage is not being used (because the .qgd file won't even exist).

**Comment by Houska1** [comment]
@pblottiere. Will test once it gets into the compiled master. Mind you, "testing" will be a long and random process, i.e. seeing whether a problem I faced irregularly about 1-2x/month will come back or not. I tried and failed a while back to find a reliable way to recreate the issue but it's sporadic even if you force a hibernation or deliberately leave a project open overnight, etc.

**Comment by jlayt** [comment]
Any chance of a backport to 3.20 and 3.16 LTR?

**Comment by pblottiere** [comment]
Hello @jlayt,

> Any chance of a backport to 3.20 and 3.16 LTR?

No, sorry, this fix is too risky for a backport.

**Comment by jlayt** [comment]
> No, sorry, this fix is too risky for a backport.

OK, thanks for considering, just glad that it's been addressed, many thanks!

**Pull request merged**